state-synchronizer/src/executor_proxy.rs

```diff
@@ -151,4 +166,58 @@ impl ExecutorProxyTrait for ExecutorProxy {
         );
         Ok(waypoint_li.clone())
     }
+
+    async fn publish_on_chain_config_updates(&mut self, events: Vec<ContractEvent>) -> Result<()> {
+        if events.is_empty() {
+            return Ok(());
+        }
+
+        let access_paths = ON_CHAIN_CONFIG_REGISTRY
+            .iter()
+            .map(|config_id| config_id.access_path())
+            .collect();
+        let curr_configs = self
+            .storage_read_client
+            .batch_fetch_config(access_paths)
+            .await;
+
+        let mut new_configs_map = HashMap::new();
+        for (id, configs) in ON_CHAIN_CONFIG_REGISTRY.iter().zip(curr_configs.iter()) {
+            new_configs_map.insert(id, configs);
+        }
+
+        // calculate deltas
+        let mut changed_configs = HashSet::new();
+        let mut new_configs = HashMap::new();
+        for ((id, config), new_config) in self.on_chain_configs.iter().zip(curr_configs) {
+            new_configs.insert(*id, new_config.clone());
+
+            // the byte array is `None` on storage read failure
+            // We should still publish reconfig notification for failed DB read, to indicate that a
+            // reconfig event did happen, regardless of DB failure
+            let mut should_publish = true;
+            if let Some(new_config) = new_config {
+                if let Some(config) = config {
+                    if config == &new_config {
+                        should_publish = false;
+                    }
+                }
+            };
+
+            if should_publish {
+                changed_configs.insert(id.clone());
+            }
+        }
+
+        let payload = OnChainConfigPayload::new(new_configs.clone());
```

Comment:
The payload probably should only hold an Arc<HashMap> to avoid copies, it may be premature optimization but seems trivial